\documentclass[9pt,twocolumn,twoside]{gsajnl}
% Use the documentclass option 'lineno' to view line numbers

\articletype{inv} % article type
% {inv} Investigation 
% {gs} Genomic Selection
% {goi} Genetics of Immunity 
% {gos} Genetics of Sex 
% {mp} Multiparental Populations

\title{Analysis of a TCGA RNA-seq data set on Colon Adenocarcinoma}

\author[$\ast$,1]{Aina Rill}
\author[$\dagger$]{Luisa Santus}
\author[$\ddagger$]{Altair C. Hernández}

\affil[$\ast$]{Master Programme on Bioinformatics for Health Sciences, Universitat Pompeu Fabra, Barcelona, Spain}
\affil[$\dagger$]{Master Programme on Bioinformatics for Health Sciences, Universitat Pompeu Fabra, Barcelona, Spain}
\affil[$\ddagger$]{Master Programme on Bioinformatics for Health Sciences, Universitat Pompeu Fabra, Barcelona, Spain}


\keywords{Colon Adenocarcinoma; RNA-seq; BioConductor; ...}

\runningtitle{Cancer Adenocarcinoma RNA-seq Analysis} % For use in the footer 

%% For the footnote.
%% Give the last name of the first author if only one author;
% \runningauthor{FirstAuthorLastname}
%% last names of both authors if there are two authors;
% \runningauthor{FirstAuthorLastname and SecondAuthorLastname}
%% last name of the first author followed by et al, if more than two authors.
%%\runningauthor{FirstAuthorLastname \textit{et al.}} is not necessary in our case.

\begin{abstract}
The abstract should be written for people who may not read the entire paper, so it must stand on its own. The impression it makes usually determines whether the reader will go on to read the article, so the abstract must be engaging, clear, and concise. In addition, the abstract may be the only part of the article that is indexed in databases, so it must accurately reflect the content of the article. A well-written abstract is the  most effective way to reach intended readers, leading to more robust search, retrieval, and usage of the article. 

Please see additional guidelines notes on preparing your abstract below.
\end{abstract}

\begin{document}

\section{Introduction}

\lettrine[lines=2]{\color{color2}C}{}olorectal Cancer (CRC) is considered to be the most common malignant cancer affecting the gastrointestinal tract, being second in males and third in females for its frequency (746,000 and 614,000 cases per year according to World Health Organization). Among the five subtypes of CRC (adenocarcinomas, carcinoid tumors, gastrointestinal stromal tumors, lymphomas and sarcomas), adenocarcinomas are the most common (95 \% of all CRCs) \citep{\textit{siegel2015cancer}}.

It is a prevalent disease in those aged up to 50, with a higher prevalence in woman, and behavioural risk factors (such as diet, alcohol intake, smoking, and physical inactivity) account for a large proportion of cases. 

Recently, several studies have been achieved in studying the molecular mechanisms of CRC formation. It can be arised from one or a combination of three differen mechanisms, namely chromosmomal inestability (CIN), CpG islands methylator phenotype (CIMP), and microsatellite inestability (MSI). 

Understanding the specific mechanisms of tumorigenesis and the underlaying genetic and epigenetic traits is crutial in the disease phenotype comprehension.

In this study we aim to evaluate the association between differential expressed genes (DE genes) and the development of the cancer. To do so, we analyze the expression profiles of patients with CRC from cohort of [The Cancer Genome Atlas](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga)(TCGA), accessible in the form of a raw RNA-seq counts produced by rahman2015alternative using a pipeline based on the R/Bioconductor software package `r Biocpkg("Rsubread")`.


\section{Materials and Methods}
\label{sec:materials:methods}

\subsection{Statistical Analysis} 

\subsubsection{Quality assessment and normalization}

For the analysis we used the BioConductor Package of R in order to correctly process and analyse tha cancer data. 

Using the CRC data set free provided by The Cancer Genome Atlas (TCGA) project (see \nameref{sec:data:availability}), a total of 524 samples (483 tumor and 41 controls) were analysed. We load the data as a \emph{SummarizedExperiment} object. For the quality assessment and normalization we follow the \textbf{\emph{edgeR}} pipeline, as is one of the most widely used.

\subsubsection{Identification of DE genes}

Differential analysis was performed for the RNA-seq data with the \emph{limma} package of \textbf{\emph{R}}. False positive Discovery rate (FDR) correction and \[log_2\]Fold Change thresholds were applied as thresholds to screen out DE genes: \[log_2\]FC > 2, and a FDR < 0.01.

\subsubsection{Functional Enrichment analysis}

Gene Set Enrichment Analysis (GSEA) method, based on the gene set level analysis, was performed to overcome DE genes. \textit{Simple GSEA} algorithm from GSVA \emph{BioConductor} package was used to download the \emph{Broad Gene Set C2 Collection} version 3.0. The analysis was focused on the following pathways: \textit{KEGG, REACTOME and BIOCARTA}. Gene sets were assessed as statistically significant by ranking according to the \textit{Z-score} statistic, with an adjusted \textit{p} < 0.01. Possibly change in scale effect in gene set analysis was evaluated by calculating the 


\subsubsection{Data Availability}
\label{sec:data:availability}

The Cancer Genome Atlas (TCGA) project (\url{https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga}) is a joint effort between the National Cancer Institute (NCI) and the National Genome Research Institute (NHCRI) to facilitate the sharing of data and speed ip cancer research. 

We used the TCGA clinical and expression data for the RNA-seq analysis, provided by the Universitat Pompeu Fabra (UPF). The datasets used are tables of counts generated by \citep{\textit{Rahman et al, 2015}} from the TCGA raw sequenced read ata using the \label{Rsubread/featureCounts} pipeline.

The code used to download the data can be accessed here: \url{https://github.com/ainarill/RNASeqLockHolmes}

\section{Results and Discussion}

The results and discussion should not be repetitive. The results section should give a factual presentation of the data and all tables and figures should be referenced; the discussion should not summarize the results but provide an interpretation of the results, and should clearly delineate between the findings of the particular study and the possible impact of those findings in a larger context. Authors are encouraged to cite recent work relevant to their interpretations. Present and discuss results only once, not in both the Results and Discussion sections. It is sometimes acceptable to combine results and discussion. The text should be as succinct as possible. Heed Strunk and White's dictum: "Omit needless words!"

\section{Additional guidelines}

\subsection{Numbers} In the text, write out numbers nine or less except as part of a date, a fraction or decimal, a percentage, or a unit of measurement. Use Arabic numbers for those larger than nine, except as the first word of a sentence; however, try to avoid starting a sentence with such a number.

\subsection{Units} Use abbreviations of the customary units of measurement only when they are preceded by a number: "3 min" but "several minutes". Write "percent" as one word, except when used with a number: "several percent" but "75\%." To indicate temperature in centigrade, use ° (for example, 37°); include a letter after the degree symbol only when some other scale is intended (for example, 45°K).

\subsection{Nomenclature and Italicization} Italicize names of organisms even when  when the species is not indicated.  Italicize the first three letters of the names of restriction enzyme cleavage sites, as in HindIII. Write the names of strains in roman except when incorporating specific genotypic designations. Italicize genotype names and symbols, including all components of alleles, but not when the name of a gene is the same as the name of an enzyme. Do not use "+" to indicate wild type. Carefully distinguish between genotype (italicized) and phenotype (not italicized) in both the writing and the symbolism.

\subsection{Cross References}
Use the \verb|\nameref| command with the \verb|\label| command to insert cross-references to section headings. For example, a \verb|\label| has been defined in the section \nameref{sec:materials:methods}.

\section{In-text Citations}

Add citations using the \verb|\citep{}| command, for example \citep{neher2013genealogies} or for multiple citations, \citep{neher2013genealogies, rodelsperger2014characterization}

\section{Examples of Article Components}
\label{sec:examples}

The sections below show examples of different header levels, which you can use in the primary sections of the manuscript (Results, Discussion, etc.) to organize your content.

\section{First level section header}

Use this level to group two or more closely related headings in a long article.

\subsection{Second level section header}

Second level section text.

\subsubsection{Third level section header:}

Third level section text. These headings may be numbered, but only when the numbers must be cited in the text. 

\section{Figures and Tables}

Figures and Tables should be labelled and referenced in the standard way using the \verb|\label{}| and \verb|\ref{}| commands.

\subsection{Sample Figure}

Figure \ref{fig:spectrum} shows an example figure.

\begin{figure}[htbp]
\centering
\includegraphics[width=\linewidth]{example-figure}
\caption{Example figure from \url{10.1534/genetics.114.173807}. Please include your figures in the manuscript for the review process. You can upload figures to Overleaf via the Project menu. Upon acceptance, we'll ask for your figure files to be uploaded in any of the following formats: TIFF (.tiff), JPEG (.jpg), Microsoft PowerPoint (.ppt), EPS (.eps), or Adobe Illustrator (.ai).  Images should be a minimum of 300 dpi in resolution and 500 dpi minimum if line art images.  RGB, CMYK, and Grayscale are all acceptable. Halftones should be high contrast with sharp detail, because some loss of detail and contrast is inevitable in the production process. Figures should be 10-20 cm in width and 1-25 cm in height. Graph axes must be exactly perpendicular and all lines of equal density.
Label multiple figure parts with A, B, etc. in bolded type, and use Arrows and numbers to draw attention to areas you want to highlight. Legends should start with a brief title and should be a self-contained description of the content of the figure that provides enough detail to fully understand the data presented. All conventional symbols used to indicate figure data points are available for typesetting; unconventional symbols should not be used. Italicize all mathematical variables (both in the figure legend and figure) , genotypes, and additional symbols that are normally italicized.  
}%
\label{fig:spectrum}
\end{figure}

\subsection{Sample Video}

Figure \ref{video:spectrum} shows how to include a video in your manuscript.

\begin{figure}[htbp]
\centering
\includegraphics[width=\linewidth]{example-figure}
\caption{Example movie (the figure file above is used as a placeholder for this example). \textit{GENETICS} supports video and movie files that can be linked from any portion of the article - including the abstract. Acceptable formats include .asf, avi, .wav, and all types of Windows Media files.   
}%
\label{video:spectrum}
\end{figure}


\subsection{Sample Table}

Table \ref{tab:shape-functions} shows an example table. Avoid shading, color type, line drawings, graphics, or other illustrations within tables. Use tables for data only; present drawings, graphics, and illustrations as separate figures. Histograms should not be used to present data that can be captured easily in text or small tables, as they take up much more space.  

Tables numbers are given in Arabic numerals. Tables should not be numbered 1A, 1B, etc., but if necessary, interior parts of the table can be labeled A, B, etc. for easy reference in the text.  


\begin{table*}[htbp]
\centering
\caption{\bf Students and their grades}
\begin{tableminipage}{\textwidth}
\begin{tabularx}{\textwidth}{XXXX}
\hline
Student & Grade\footnote{This is an example of a footnote in a table. Lowercase, superscript italic letters (a, b, c, etc.) are used by default. You can also use *, **, and *** to indicate conventional levels of statistical significance, explained below the table.} & Rank & Notes \\
\hline
Alice & 82\% & 1 & Performed very well.\\
Bob & 65\% & 3 & Not up to his usual standard.\\
Charlie & 73\% & 2 & A good attempt.\\
\hline
\end{tabularx}
  \label{tab:shape-functions}
\end{tableminipage}
\end{table*}

\section{Sample Equation}

Let $X_1, X_2, \ldots, X_n$ be a sequence of independent and identically distributed random variables with $\text{E}[X_i] = \mu$ and $\text{Var}[X_i] = \sigma^2 < \infty$, and let
\begin{equation}
S_n = \frac{X_1 + X_2 + \cdots + X_n}{n}
      = \frac{1}{n}\sum_{i}^{n} X_i
\label{eq:refname1}
\end{equation}
denote their mean. Then as $n$ approaches infinity, the random variables $\sqrt{n}(S_n - \mu)$ converge in distribution to a normal $\mathcal{N}(0, \sigma^2)$.

\bibliography{example-bibliography}




\end{document}