---
output:
  BiocStyle::html_document
---

<!---
Because we split the analysis pipeline in different independent files,
to speed up processing it, here in the setup block we load libraries and
objects that were loaded or produced in the previously processed file,
and which are necessary in this file.
--->



```{r setup, cache=FALSE, echo=FALSE, results='hide', message=FALSE}
library(knitr)
library(SummarizedExperiment)
library(edgeR)
library(geneplotter)
library(sva)
library(limma)
library(calibrate)
library(EnhancedVolcano)

opts_chunk$set(cache=TRUE,
               cache.path="cache/file2",
               cache.extra=R.version.string,
               autodep=TRUE,
               fig.align="center",
               comment="")

coadse.tumor.filt <- readRDS(file.path("results", "coadse.tumor.filt.rds"))
dge.tumor.filt <- readRDS(file.path("results", "dge.tumor.filt.rds"))

```

# Differential expression between tumor stages

Once we have normalized the data from the 4th tumor stages, we are interested in identifying gene expression changes and associated p-values between those tumor stages. For this task, we used the R/Bioconductor packages [sva](http://bioconductor.org/packages/sva) and [limma](https://bioconductor.org/packages/release/bioc/html/limma.html).

# Model 1 (SIGUIENDO PIPELINE DE_GENES_II)
In the follwing step we create the design matrix. The outcome of interest will be the different stages, so the first step is to build the design matrix. 
```{r}
mod <- model.matrix(~ajcc_pathologic_tumor_stage , colData(coadse.tumor.filt))
head(coadse.tumor.filt$gender)
head(mod, n=5)
```
Next, we want to adjust for the mean-variance realtionship. We proceed with the voom() function, since as we can observe in the Figure \@ref(fig:libsize), there are big differences in sample sizes among the samples. 
```{r libsize}
barplot(sort(dge.tumor.filt$samples$lib.size)/1e+06, xlab = "Samples", ylab = "Library sizes (Millions)")
```

So, we calculate the weights that estimate the mean-variance realtionship at gene-by-sample level with the voom() function and we will use these weights to fit the model. 
```{r}
v <- voom(dge.tumor.filt, mod, plot=TRUE)
```

Next, we fit the model with the `lmFit` method.
```{r}
fit<- lmFit(v,mod)
```
We then calculate the moderated t-statistics with `eBayes`.
```{r}
fit <- eBayes(fit)
FDRcutoff <- 0.01
```

Here, we examine the differential expression analysis result of this part at a `r FDRcutoff` to have a broad overview.
```{r}
#Classify a series of related t-statistics as up, down or not significant. 
res1 <- decideTests(fit, p.value = 0.01) 
summary(res1)
```
With an FDR cutoff of 0.01 we can notice that there are not up or down results for the different stages. 
AQUÍ PODEMOS VER QUE NO HAY RESULTADOS SIGNIFICATIVOS PARA EL T-TEST, PARA UN CUTOFF DEL 0.01. NO OBSTANTE, SÍ PARA 0.05.
```{r}
#Classify a series of related t-statistics as up, down or not significant. 
res2 <- decideTests(fit, p.value = 0.05) 
summary(res2)
```
SEGUIMOS CON FDR < 0.05!

We extract the genes chromosome to look at the distribution of the differential expressed genes among the chromosomes, again for the purpose of getting an overview of the data we will be working with.
```{r}
genesmd <- data.frame(chr = as.character(sub("\\_.*", "", seqnames(rowRanges(coadse.tumor.filt)))), symbol = rowData(coadse.tumor.filt)[, 1], stringsAsFactors = FALSE)
#Extract a table of the top-ranked genes from a linear model fit.
tt <- topTable(fit, coef = 2, n = Inf)
fit$genes <- genesmd
#head(tt)
sort(table(tt$chr[tt$adj.P.Value < FDRcutoff]), decreasing = TRUE) # adj.P.val is higher than p value
```
TAMPOCO ENCUENTRA RESULTADOS SIGNIFICATIVOS.

```{r}
#DEgenes <- rownames(tt)[tt$P.Val < FDRcutoff]
#length(DEgenes)
```

```{r}
EnhancedVolcano(tt,
    lab = tt$symbol,
    x = 'logFC',
    y = 'P.Value',
    xlim = c(-5, 5),
    title = 'Volcano Plot',
    pCutoff = 10e-16,
    FCcutoff = 3,
    transcriptPointSize = 1.0,
    transcriptLabSize = 3.0)
```

##------------------------------------------
# Differential expression - model 2

LO PRUEBO CONSTRUYENDO UNA m0:

```{r}
mod2 <- model.matrix(~ajcc_pathologic_tumor_stage, colData(coadse.tumor.filt))
mod0 <- model.matrix(~1, colData(coadse.tumor.filt)) # NINGÚN VALOR PARA AJUSTAR
#Verify wether is full rank --> yes!
is.fullrank(mod)
pv <- f.pvalue(assays(coadse.tumor.filt)$logCPM, mod2, mod0)
sum(p.adjust(pv, method="fdr") < 0.01)
```
There are `r sum(p.adjust(pv, method="fdr") < 0.01)` genes changing significantly their expression at FDR < 1%. In Figure \@ref(fig:pdist) below we show the distribution of the resulting p-values.
The distribution of the p-value looks uniform with a peak a low p-values, indicating we are not under the null hypothesis of no differential expression but we are actually having differential expression values between tumor and control samples.
```{r pdist, echo=FALSE, out.width="400px", fig.cap="Distribution of raw p-values for an F-test on every gene between tumor and normal samples."}
hist(pv, main="", las=1, xlab="P-value")
```
Then we are going to estimate surrogate variables using the `sva()` function.
```{r}
sv <- sva(assays(coadse.tumor.filt)$logCPM, mod2, mod0)
sv$n
```
The SVA algorithm has found `r sv$n` surrogate variables. So, we are going to use them to
assess againt the extent of differential expression this time adjusting for these
surrogate variables.

```{r}
modsv <- cbind(mod2, sv$sv)
mod0sv <- cbind(mod0, sv$sv)
pvsv <- f.pvalue(assays(coadse.tumor.filt)$logCPM, modsv, mod0sv)
sum(p.adjust(pvsv, method="fdr") < 0.01)
```
AL AJUSTAR NO INCREMENTA EL NUMERO DE GENES. EN EL HISTOGRAMA PARECE QUE LOS LOW P VALUES BAJAN CUANDO AJUSTAMOS POR SVA. NO ENTIENDO PORQUÉ.

```{r psvdist, echo=FALSE, out.width="400px", fig.cap="Distribution of raw p-values for an F-test on every gene between tumor and normal samples, adjusting for surrogate variables estimated with SVA."}
hist(pvsv, main="", las=1, xlab = "P-value")
```
Next, we want to adjust for the mean-variance realtionship. Two main methods are currently available. We decide to proceed with the voom() function, since as we can observe in the Figure \@ref(fig:libsize), there are big differences in sample sizes among the samples. 
```{r libsize}
barplot(sort(dge.tumor.filt$samples$lib.size)/1e+06, xlab = "Samples", ylab = "Library sizes (Millions)")
```

So, we calculate the weights that estimate the mean-variance realtionship at gene-by-sample level with the voom() function and we will use these weights to fit the model. 
```{r}
v2 <- voom(dge.tumor.filt, modsv, plot=TRUE)
```
Next, we fit the model. 
```{r}
fit2<- lmFit(v2,modsv)
```
We then calculate the moderated t-statistics.
```{r}
# compute moderated t-statistics, moderated F-statistic, and log-odds of differential expression by empirical Bayes moderation of the standard errors towards a common value.
fit2 <- eBayes(fit2)
FDRcutoff <- 0.01
```
Here, we examine the differential expression analysis result of this part at a `r FDRcutoff` to have a broad overview.
```{r}
#Classify a series of related t-statistics as up, down or not significant. 
res2 <- decideTests(fit, p.value = FDRcutoff) 
summary(res2)
```
PASA LO MISMO QUE ANTES.
We extract the genes chromosome to look at the distribution of the differential expressed genes among the chromosomes, again for the purpose of getting an overview of the data we will be working with.
```{r}
genesmd2 <- data.frame(chr = as.character(sub("\\_.*", "", seqnames(rowRanges(coadse.tumor.filt)))), symbol = rowData(coadse.tumor.filt)[, 1], stringsAsFactors = FALSE)
unique(genesmd$chr)
#Extract a table of the top-ranked genes from a linear model fit.
tt2 <- topTable(fit2, coef = 2, n = Inf)
head(tt2)
fit2$genes <- genesmd2
sort(table(tt2$chr[tt2$adj.P.Val < FDRcutoff]), decreasing = TRUE)
```


```{r}
DEgenes <- rownames(tt2)[tt2$adj.P.Val < FDRcutoff]
length(DEgenes)
```
NINGÚN GEN PASA EL FILTRO DE P VALUE ADJUSTED!! 
```{r}
EnhancedVolcano(tt,
    lab = tt$symbol,
    x = 'logFC',
    y = 'P.Value',
    xlim = c(-5, 5),
    title = 'Volcano Plot',
    pCutoff = 10e-16,
    FCcutoff = 3,
    transcriptPointSize = 1.0,
    transcriptLabSize = 3.0)
```
## Session information

```{r, message=FALSE}
sessionInfo()
```
